src/pybind/mgr/progress/module.py

```diff
@@ -615,3 +658,12 @@ def handle_command(self, _, cmd):
             return 0, json.dumps(self._json(), indent=2), ""
         else:
             raise NotImplementedError(cmd['prefix'])
+    
+    # For other module to use.
+    def construct_pgid(self, pool_id, ps):
+        return PgId(pool_id, ps)
+
+    def construct_pg_recovery_event(self, message, refs, which_pgs, evacuate_osds, start_epoch):
+        return PgRecoveryEvent(message, refs, which_pgs,
```

Comment:
probably you forgot to update `PgRecoveryEvent` 's constructor?
```
2019-08-06T05:54:17.649 INFO:tasks.ceph.mgr.x.mira105.stderr:2019-08-06T05:54:17.673+0000 7f9db202b700 -1 Traceback (most recent call last):
2019-08-06T05:54:17.649 INFO:tasks.ceph.mgr.x.mira105.stderr:  File "/usr/share/ceph/mgr/balancer/module.py", line 626, in serve
2019-08-06T05:54:17.649 INFO:tasks.ceph.mgr.x.mira105.stderr:    self.execute(plan)
2019-08-06T05:54:17.650 INFO:tasks.ceph.mgr.x.mira105.stderr:  File "/usr/share/ceph/mgr/balancer/module.py", line 1308, in execute
2019-08-06T05:54:17.650 INFO:tasks.ceph.mgr.x.mira105.stderr:    old_dump, new_osdmap, plan.message)
2019-08-06T05:54:17.650 INFO:tasks.ceph.mgr.x.mira105.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1273, in remote
2019-08-06T05:54:17.650 INFO:tasks.ceph.mgr.x.mira105.stderr:    args, kwargs)
2019-08-06T05:54:17.650 INFO:tasks.ceph.mgr.x.mira105.stderr:RuntimeError: Remote method threw exception: Traceback (most recent call last):
2019-08-06T05:54:17.650 INFO:tasks.ceph.mgr.x.mira105.stderr:  File "/usr/share/ceph/mgr/progress/module.py", line 403, in balancer_crush_compat
2019-08-06T05:54:17.651 INFO:tasks.ceph.mgr.x.mira105.stderr:    start_epoch=self.get_osdmap().get_epoch()
2019-08-06T05:54:17.651 INFO:tasks.ceph.mgr.x.mira105.stderr:TypeError: __init__() got an unexpected keyword argument 'evacuate_osds'
2019-08-06T05:54:17.651 INFO:tasks.ceph.mgr.x.mira105.stderr:
```

see http://pulpito.ceph.com/kchai-2019-08-06_05:23:57-rados-wip-kefu-testing-2019-08-06-1102-distro-basic-mira/4187832/

Reply:
Yep I suspect that because the master branch has "which_osds" as the var name instead of evacuate_osds.